feat(suite): disable coinjoin resume if something wrong

karliatto opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

User shouldn't be able to resume CoinJoin if there is some problem

    * Backend disconnected
    * No internet connection
    * No Tor connection (it was already implemented)
    * Device disconnected
    
## Related Issue

Resolve https://github.com/trezor/trezor-suite/issues/6908

karliatto commented on packages/suite/src/utils/wallet/accountUtils.ts:
```diff
@@ -25,3 +25,14 @@ export const getSelectedAccount = (
     );
 };
+
+export const getIsAccountDisconnected = (
+    blockchain: BlockchainNetworks,
+    selectedAccount: SelectedAccountStatus,
+) => {
+    if (selectedAccount.status !== 'loaded') return true;
+    const { symbol } = selectedAccount.network;
+    const chain = blockchain[symbol];
+    if (!chain || chain.connected) return null;
+    return !chain || !chain.connected;
+};
```
@marekrjpolak I got this as inspiration from https://github.com/trezor/trezor-suite/blob/develop/packages/suite/src/components/wallet/AccountMode/BackendDisconnected.tsx#L61

It looks like after disconnecting (for example disconnecting computer form internet) it takes a while to realize that backend is not connected. I was wondering if you know about better way to get that information?

marekrjpolak replied:
Here the `chain.connected` doesn't work because coinjoin account is using its own non-standard backend instead of default TrezorConnect backends (meaning that you could have `chain.connected = true` even if CJ backend is inaccessible). Consider using `account.status === 'out-of-sync'` instead.

Still, it updates only once in a minute currently, but we could change it if needed.

karliatto replied:
Thanks for the suggestion, I updated it to use `account.status === 'out-of-sync'`  [9720a5b](https://github.com/trezor/trezor-suite/pull/7295/commits/9720a5bafe8e2e578169e4b76882f051f84a1e28)

marekrjpolak commented on packages/suite/src/components/wallet/CoinjoinSummary/CoinjoinStatus.tsx:
```diff
@@ -152,12 +155,38 @@ interface CoinjoinStatusProps {
+        coinjoinBlockedMessageId = 'TR_UNAVAILABLE_COINJOIN_DEVICE_DISCONNECTED';
+    } else if (isAccountDisconnected) {
+        coinjoinBlockedMessageId = 'TR_UNAVAILABLE_COINJOIN_ACCOUNT_DISCONNECTED';
+    } else if (!online) {
```
I would change the priorities of conditions. If you're not `online`, account should be in theory always disconnected so the `Unavailable. No internet connection.` message would never be shown. And maybe extract the coinjoin-blocked logic into a separate hook? It could be a bit clearer.

karliatto replied:
That makes a lot of sense. I moved it to a new hook and updated the priorities of condition so it is more reasonable. [9720a5b](https://github.com/trezor/trezor-suite/pull/7295/commits/9720a5bafe8e2e578169e4b76882f051f84a1e28)

karliatto commented on packages/suite/src/components/wallet/CoinjoinSummary/CoinjoinStatus.tsx:
```diff
@@ -270,9 +269,9 @@ export const CoinjoinStatus = ({ session, accountKey }: CoinjoinStatusProps) =>
             );
         }
-        if (isCoinJoinBlockedByTor) {
+        if (isCoinjoinResumeBlocked) {
```
@marekrjpolak I rebased it with your latest changes including Feature Flag. It still looks to me that feature flag should has different `Tooltip` section separated from the rest of the  reasons that are blocking resume coinjoin.

dahaca commented:
I completely missed this PR and implemented the check for a device in [another PR](https://github.com/trezor/trezor-suite/pull/7345) 🫠 I also refactored the logic slightly and added a background change for the counter when there's a blocker. One of us could merge first, so all the best parts are included, what do you say to that? :)

karliatto commented:
> I completely missed this PR and implemented the check for a device in [another PR](https://github.com/trezor/trezor-suite/pull/7345) 🫠 I also refactored the logic slightly and added a background change for the counter when there's a blocker. One of us could merge first, so all the best parts are included, what do you say to that? :)

@dahaca sure I agree, why don't we merge first this one since it was created first?

dahaca commented on packages/suite/src/hooks/coinjoin/useBlockedCoinjoinResume.ts:
```diff
@@ -0,0 +1,39 @@
+    }));
+    const isAccountOutOfSync = getIsCoinjoinOutOfSync(selectedAccount);
+    const deviceStatus = useSelector(selectDeviceState);
+    const isDeviceDisconnected = deviceStatus !== 'connected';
```
In mine I was checking the `available` flag, in this case they are the same I suppose?
